### PR TITLE
fix: permission error when cleaning up fetches directory in run-task

### PR DESCRIPTION
**Pull request by gabrielBusta**
When we run VPN tasks on [win2012 workers](https://github.com/mozilla-mobile/mozilla-vpn-client/blob/main/taskcluster/ci/config.yml#L40) they run successfully but they fail to clean up fetched content.

Here's one of these [tasks](https://firefox-ci-tc.services.mozilla.com/tasks/KcekQk81SPuSzG9obKpkMw/runs/0/logs/public/logs/live.log).

Here's the error:
```
Traceback (most recent call last):
  File "run-task", line 1118, in <module>
    sys.exit(main(sys.argv[1:]))
  File "run-task", line 1113, in main
    shutil.rmtree(fetches_dir)
  File "C:\mozilla-build\python3\lib\shutil.py", line 494, in rmtree
    return _rmtree_unsafe(path, onerror)
  File "C:\mozilla-build\python3\lib\shutil.py", line 384, in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
  File "C:\mozilla-build\python3\lib\shutil.py", line 384, in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
  File "C:\mozilla-build\python3\lib\shutil.py", line 384, in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
  [Previous line repeated 4 more times]
  File "C:\mozilla-build\python3\lib\shutil.py", line 389, in _rmtree_unsafe
    onerror(os.unlink, fullname, sys.exc_info())
  File "C:\mozilla-build\python3\lib\shutil.py", line 387, in _rmtree_unsafe
    os.unlink(fullname)
PermissionError: [WinError 5] Access is denied: 'Z:\\task_165098658390600\\fetches\\VisualStudio\\VC\\Tools\\MSVC\\14.30.30705\\bin\\Hostx64\\x64\\Microsoft.Diagnostics.Tracing.EventSource.dll'
```

I think this code is failing because `run-task` doesn't have the access to clean up the `fetches` directory.
This is a patch where we make sure we have the access we need to the files before we try to delete them.

### CODE REVIEW

**Review by hneiva — APPROVED** [review]
LGTM

**Review by ahal — APPROVED** [review]
Neat, I didn't know about this `onerror` callback!

**Comment by gabrielBusta** [comment]
@ahal @hneiva turns out there's code in Mozilla Central that accounts for issues like these. I grabbed this code from [mozfile.py](https://searchfox.org/mozilla-central/source/testing/mozbase/mozfile/mozfile/mozfile.py#187)

What do you think? I translated some of the unit test too. There's a couple more test cases on Mozilla Central but they'd only make sense if we ran these tests in NT/Windows (I don't think we do that?)

**Review by hneiva — CHANGES_REQUESTED** [review]

File: src/taskgraph/run-task/run-task [hunk 1]
```diff
@@ -868,6 +868,118 @@ def maybe_run_resource_monitoring():
...
+        try:
+            func(*args)
+        except OSError as e:
+            # Error codes are defined in:
+            # http://docs.python.org/2/library/errno.html#module-errno
+            if e.errno not in (errno.EACCES, errno.ENOTEMPTY):
+                raise
+
+            if retry_count == retry_max:
+                raise
```
[the remark this answers is not in the document]
Comment (jcristau) [reply src/taskgraph/run-task/run-task:890]: They're basically the same thing, no?

File: src/taskgraph/run-task/run-task [hunk 2]
```diff
@@ -868,6 +868,118 @@ def maybe_run_resource_monitoring():
...
+    retry_delay is multiplied by the number of failed attempts to increase
+    the likelihood of success in subsequent attempts.
+    """
+    retry_count = 0
+    while True:
+        try:
+            func(*args)
+        except OSError as e:
+            # Error codes are defined in:
+            # http://docs.python.org/2/library/errno.html#module-errno
```
Comment (jcristau) [inline src/taskgraph/run-task/run-task:885]: maybe s/2/3/

**Review by ahal — CHANGES_REQUESTED** [review]
Thanks, looks good!

Just a few minor nits from myself and others, enough to warrant a re-review I think.

**Comment by ahal** [comment]
Also if you rebase on latest `main` you should get the proper codecov delta :)

File: src/taskgraph/run-task/run-task [hunk 3]
```diff
@@ -1110,7 +1222,10 @@ def main(args):
         fetches_dir = os.environ.get('MOZ_FETCHES_DIR')
         if fetches_dir and os.path.isdir(fetches_dir):
             print_line(b'fetches', b'removing %s\n' % fetches_dir.encode('utf-8'))
-            shutil.rmtree(fetches_dir)
+            if IS_WINDOWS:
+                remove(fetches_dir)
+            else:
+                shutil.rmtree(fetches_dir)
```
[the remark this answers is not in the document]
Comment (gabrielBusta) [reply src/taskgraph/run-task/run-task]: Didn't change it at all. Calling on non-windows as well now

File: src/taskgraph/run-task/run-task [hunk 4]
```diff
@@ -868,6 +868,118 @@ def maybe_run_resource_monitoring():
...
+    """
+
+    import shutil
+
+    def _call_with_windows_retry(*args, **kwargs):
+        try:
+            _call_windows_retry(*args, **kwargs)
+        except OSError as e:
+            # The file or directory to be removed doesn't exist anymore
+            if e.errno != errno.ENOENT:
```
[the remark this answers is not in the document]
Comment (gabrielBusta) [reply src/taskgraph/run-task/run-task]: [This cleaned up nicely :)](https://github.com/taskcluster/taskgraph/commit/66099eed4d7883ad383e7768ebc4f5dc963fe192)

File: src/taskgraph/run-task/run-task (hunk 1 above)
[the remark this answers is not in the document]
Comment (gabrielBusta) [reply src/taskgraph/run-task/run-task:890]: It is equivalent... *almost*. The stack trace looks a little diff. For `raise e` the tack trace contains the `raise e` line. For `raise` the stack trace stops at the line that actually triggered the exception. There's also extra magic in the run time that figures out the "active exception being handled". In Python **2** `raise` is better that `raise e`. With `raise` the correct line that caused the exception is quoted in the stack trace (in **2** `raise e` actually makes it so the original source of the exception is not in the stack trace)

From the [spec](https://docs.python.org/3/reference/simple_stmts.html?highlight=raise#the-raise-statement):

> If no expressions are present, raise re-raises the exception that is currently being handled, which is also known as the active exception. If there isn’t currently an active exception, a RuntimeError exception is raised indicating that this is an error

**Comment by gabrielBusta** [comment]
@ahal @hneiva I noticed from looking at @strseb PR that there's also issues running `git clean` on these windows workers. It's failing with the same permission error. [We could delete the files with the sturdier remove function](https://github.com/taskcluster/taskgraph/compare/fetches-cleanup-permission-error-patch...fetches-cleanup-permission-error-patch-git-clean-issue?expand=1#diff-e29e172140c458b95d3f5b1a5542188ffe334182ccef1c3b62e1eeb102beec1aR521). Why haven't we run into these before? I guess the gecko builds use HG? Any git based-projects building on Windows?

**Comment by gabrielBusta** [comment]
@ahal in M-C there's no `fetches_dir` clean up and there's no `git clean` because it's tailored to `hg`

**Review by ahal — APPROVED** [review]
Awesome, lgtm!

**Pull request merged**